File: usaco/training/Section 3.2/butter.py
"""
ID: jgwilso1
LANG: PYTHON3
TASK: butter
"""
import heapq

fin = open ('butter.in', 'r')
fout = open ('butter.out', 'w')
N, P, C = map(int, fin.readline().split())

# store in table number of cows in each pasture
ncow = [0 for _ in range(P + 1)]
for _ in range(N):
    p = int(fin.readline())
    ncow[p] += 1

# create graph using adjacency list
adj = [[] for _ in range(P + 1)]
for _ in range(C):
    a, b, dist = map(int, fin.readline().split()) 
    adj[a].append((b, dist))
    adj[b].append((a, dist))

# shortest path every pasture w/ dijkstra's
res = 1e9
for source in range(1, P + 1):
    pq = []
    shortest_path = [1e9 for _ in range(P + 1)]
    processed = [False for _ in range(P + 1)]
    heapq.heappush(pq, (0, source))
    shortest_path[source] = 0
    while len(pq) > 0:
        top = heapq.heappop(pq)
        u = top[1]
        if processed[u] is True:
            continue
        processed[u] = True
        for neighbor in adj[u]:
            v = neighbor[0]
            d = neighbor[1]
            if shortest_path[u] + d < shortest_path[v]:
                shortest_path[v] = shortest_path[u] + d
                heapq.heappush(pq, (shortest_path[v], v))
    cnt = 0
    for i in range(1, P + 1):
        cnt += shortest_path[i] * ncow[i]
    res = min(res, cnt)

fout.write (str(res) + '\n')
fin.close()
fout.close()

File: usaco/training/Section 3.2/butter.cpp
/*
ID: jgwilso1
PROG: butter
LANG: C++                 
*/
#include <iostream>
#include <cstdio>
#include <cassert>
#include <algorithm>
#include <queue>
#include <utility>
#include <vector>
using namespace std;

const int INF = 1e9;
int cow[501];
vector<pair<int, int>> adj[801];

int main() {
  freopen("butter.in", "r", stdin);
  freopen("butter.out", "w", stdout);
  int N, P, C; // cows, pastures, connecting paths
  cin >> N >> P >> C;
  for (int i = 1; i <= N; i++) {
    cin >> cow[i];
  }
  for (int i = 1; i <= C; i++) {
    int a, b, d;
    cin >> a >> b >> d;
    adj[a].push_back({b, d});
    adj[b].push_back({a, d});
  }
  // dijkstra greedy shortest path each pasture
  int res = INF;
  priority_queue<pair<int, int>> mn_heap; // contains {cur dist to, node}, easily access min dist node
  for (int source = 1; source <= P; source++) {
    int dist[P + 1];
    bool processed[P + 1];
    for (int i = 1; i <= P; i++) {
      dist[i] = INF;
      processed[i] = false;
    }
    dist[source] = 0;
    mn_heap.push({0, source});
    while (!mn_heap.empty()) {
      int cur = mn_heap.top().second;
      mn_heap.pop();
      if (processed[cur]) 
        continue;
      processed[cur] = true;
      for (auto neighbor : adj[cur]) {
        int n = neighbor.first;
        int w = neighbor.second;
        if (dist[cur] + w < dist[n]) {
          dist[n] = dist[cur] + w;
          mn_heap.push({-dist[n], n}); // negative hack min elements
        }
      }
    }
    int c = 0;
    for (int i = 1; i <= N; i++) {
      c += dist[cow[i]];
    }
    res = min(res, c);
  }
  cout << res << endl;

  return 0;
}
